odl/deform/mass_preserving.py

```diff
@@ -0,0 +1,548 @@
+from builtins import super
+
+from odl.discr import DiscreteLp, DiscreteLpVector
+from odl.operator.operator import Operator
+from odl.space import ProductSpaceVector, ProductSpace
+
+
+__all__ = ('GeneralDeformFixedTempl', 'GeneralDeformFixedDeform',
+           'MassPresvDeformFixedTempl', 'MassPresvDeformFixedDeform')
+
+
+def _general_deform(template, deformation, out=None):
+    """Deform a template with a general deformation field.
+
+    The function maps a given template ``I`` and a given deformation
+    field ``phi`` to the new function ``x --> I(phi(x))``.
+
+    Parameters
+    ----------
+    template : `DiscreteLpVector`
+        Template to be deformed by a deformation field.
+    deformation : element of power space of `DiscreteLp`
+        The deformation field used to deform the template.
+    out : `numpy.ndarray`, optional
+        Array to which the function values of the deformed template
+        are written. It must have the same shape as ``template`` and
+        a data type compatible with ``template.dtype``.
+
+    Returns
+    -------
+    deformed_template : `numpy.ndarray`
+        Function values of the deformed template. If ``out`` was given,
+        the returned object is a reference to it.
+
+    Examples
+    --------
+    Create a simple 1D template and a deformation field. Where the deformation
+    is identity, the output value is the same as the input value.
+    In the 4-th point, the value is taken from 0.2 (one cell) to
+    the left, i.e. 1.0.
+
+    >>> import odl
+    >>> space = odl.uniform_discr(0, 1, 5, interp='nearest')
+    >>> deform_field_space = space.vector_field_space
+    >>> template = space.element([0, 0, 1, 0, 0])
+    >>> deform_field = deform_field_space.element([[0.1, 0.3, 0.5, 0.5, 0.9]])
+    >>> _general_deform(template, deform_field)
+    array([ 0.,  0.,  1.,  1.,  0.])
+
+    The result depends on the chosen interpolation. With 'linear'
+    interpolation and an offset equal to half the distance between two
+    points, 0.1, one gets the mean of the values.
+
+    >>> space = odl.uniform_discr(0, 1, 5, interp='linear')
+    >>> deform_field_space = space.vector_field_space
+    >>> template = space.element([0, 0, 1, 0, 0])
+    >>> deform_field = deform_field_space.element([[0.1, 0.3, 0.5, 0.6, 0.9]])
+    >>> _general_deform(template, deform_field)
+    array([ 0. ,  0. ,  1. ,  0.5,  0. ])
+    """
+    image_pts = template.space.points()
+    for i, vi in enumerate(deformation):
+        image_pts[:, i] = vi.ntuple.asarray()
+    return template.interpolation(image_pts.T, out=out, bounds_check=False)
+
+
+def _mass_presv_deform(template, mass_presv_dfield, out=None):
+    """Deform a template with a mass-preserving deformation field.
+
+    The function maps a given template ``I``, a mass-preserving deformation
+    field, i.e., a given deformation field ``phi`` and its Jacobian
+    determinant to the new function ``x --> |Dphi(x)|I(phi(x))``.
```

Comment:
I don't remember the notion "deformation field" ever being used. To me, it should either be "deformation" or "displacement field".